lttng-kallsyms.c

```diff
@@ -54,6 +56,104 @@ int lttng_enumerate_kernel_symbols(struct lttng_session *session)
 }
 EXPORT_SYMBOL_GPL(lttng_enumerate_kernel_symbols);
 
+#ifdef CONFIG_MODULES
+
+/*
+ * Trace the symbols coming from a specific module
+ *
+ * data: The new module for which we want the symbols
+ * symbol_name: The function name this symbol resolves to
+ * module: The module containing this symbol. Can be NULL
+ * symbol_addr: The symbol address
+ */
+static
+int _lttng_trace_module_symbol(void * data, const char * symbol_name,
+		struct module * module, unsigned long symbol_addr)
+{
+	struct module *mod = data;
+
+	// Trace the symbols from the new module
+	if (mod == module) {
+		trace_lttng_kallsyms_new_module_symbol(symbol_addr, symbol_name, mod->name);
+	}
+	return 0;
+}
+
+static
+int lttng_kallsyms_module_coming(struct module *mod)
+{
+	int ret = 0;
+
+	ret = kallsyms_on_each_symbol(_lttng_trace_module_symbol, (void *) mod);
```

Comment:
Can you replace this iteration on the entire kernel by a more specific use of struct module from module.h ?

Reply:
Something like a copy-paste of the implementation of the module_kallsyms_on_each_symbol from module.c? Yes I can, I guess that's not something that changes  a lot in the kernel, so not much of a pain to maintain?

Reply:
If you copy-paste code from the Linux kernel, you need to state it clearly in the changelog, and make sure the file into which it is copied has a GPLv2 license header (*not* the "or LGPL-2.1").

It would make sense to have a GPLv2 only license for the lttng-kallsyms.c file, since I don't plan on sharing that code with lttng-ust.

Reply:
and indeed, I don't expect that part of struct module to change much, so not too much pain to maintain.

Reply:
by the way, grabbing the modules_mutex might be a problem within a module coming/going notifier, so it's really better to just focus on the current struct module.